File: app/src/main/java/com/byd/dashcast/voice/VoiceCommandReceiver.java
package com.byd.dashcast.voice;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import com.byd.dashcast.AppLogger;
import com.byd.dashcast.ui.diag.DiagActivity;
import com.byd.dashcast.ui.log.LogActivity;

/**
 * Receives {@link VoiceCommandRouter#ACTION_VOICE_COMMAND} local broadcasts and
 * translates each command string into the appropriate Host action.
 *
 * Register / unregister via LocalBroadcastManager in onStart / onStop.
 */
public final class VoiceCommandReceiver extends BroadcastReceiver {

    private static final String TAG = "VoiceCommandReceiver";

    public interface Host {
        boolean isActivityAlive();
        void activateCluster();
        void restoreBydDashboard();
        void startActivity(Intent intent);
        void quickSwitchToApp(String pkg);
    }

    private final Host mHost;

    public VoiceCommandReceiver(Host host) {
        mHost = host;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (!mHost.isActivityAlive() || intent == null) return;
        String cmd = intent.getStringExtra(VoiceCommandRouter.EXTRA_CMD);
        if (cmd == null) return;

        if (VoiceCommandRouter.CMD_CLUSTER_ON.equals(cmd)) {
            mHost.activateCluster();
        } else if (VoiceCommandRouter.CMD_CLUSTER_OFF.equals(cmd)) {
            mHost.restoreBydDashboard();
        } else if (VoiceCommandRouter.CMD_OPEN_DIAG.equals(cmd)) {
            mHost.startActivity(new Intent(context, DiagActivity.class));
        } else if (VoiceCommandRouter.CMD_OPEN_LOGS.equals(cmd)) {
            mHost.startActivity(new Intent(context, LogActivity.class));
        } else if (VoiceCommandRouter.CMD_LAUNCH_ON_CLUSTER.equals(cmd)) {
            String pkg = intent.getStringExtra(VoiceCommandRouter.EXTRA_PKG);
            if (pkg != null) mHost.quickSwitchToApp(pkg);
        } else {
            AppLogger.d(TAG, "Voice: unhandled cmd=" + cmd);
        }
    }
}

File: app/src/main/java/com/byd/dashcast/voice/VoiceCommandReceiver.kt
package com.byd.dashcast.voice

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent

import com.byd.dashcast.util.AppLogger
import com.byd.dashcast.ui.diag.DiagActivity
import com.byd.dashcast.ui.log.LogActivity

/**
 * Receives [VoiceCommandRouter.ACTION_VOICE_COMMAND] local broadcasts and
 * translates each command string into the appropriate [Host] action.
 *
 * Register / unregister via LocalBroadcastManager in onStart / onStop.
 */
class VoiceCommandReceiver(private val mHost: Host) : BroadcastReceiver() {

    /** Callback surface implemented by the hosting Activity (see MainActivity). */
    interface Host {
        fun isActivityAlive(): Boolean
        fun activateCluster()
        fun restoreBydDashboard()
        fun startActivity(intent: Intent)
        fun quickSwitchToApp(pkg: String)
    }

    // intent is nullable: the platform can deliver a null Intent (the Java guarded it).
    // context is never null-checked in the original and the framework always supplies one.
    override fun onReceive(context: Context, intent: Intent?) {
        if (!mHost.isActivityAlive() || intent == null) return
        val cmd = intent.getStringExtra(VoiceCommandRouter.EXTRA_CMD) ?: return

        when (cmd) {
            VoiceCommandRouter.CMD_CLUSTER_ON  -> mHost.activateCluster()
            VoiceCommandRouter.CMD_CLUSTER_OFF -> mHost.restoreBydDashboard()
            VoiceCommandRouter.CMD_OPEN_DIAG   -> mHost.startActivity(Intent(context, DiagActivity::class.java))
            VoiceCommandRouter.CMD_OPEN_LOGS   -> mHost.startActivity(Intent(context, LogActivity::class.java))
            VoiceCommandRouter.CMD_LAUNCH_ON_CLUSTER ->
                intent.getStringExtra(VoiceCommandRouter.EXTRA_PKG)?.let { mHost.quickSwitchToApp(it) }
            else -> AppLogger.d(TAG, "Voice: unhandled cmd=$cmd")
        }
    }

    companion object {
        private const val TAG = "VoiceCommandReceiver"
    }
}
